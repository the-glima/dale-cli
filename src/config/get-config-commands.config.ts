import chalk from 'chalk'

import {ConfigCommand, configCommandsMap} from '../models/config-commands.model'
import {settings} from '../settings'
import {findKey} from '../utils/find-key.util'

import {getConfigFile} from './get-config-file.config'

const isConfigCommandsValid = (commands: any): boolean => {
  return Object.keys(commands).every((cmd: string) => configCommandsMap.includes(cmd as ConfigCommand))
}

const getConfigCommands = (config?: any): string[] => {
  config = config || getConfigFile(config)
  const defaultCommands = findKey('commands', getConfigFile(settings.configPath.default))
  let commands = findKey('commands', config)

  if (commands && !isConfigCommandsValid(commands)) {
    throw new Error(
      `Invalid command in commands configuration, it should have only the following: ${chalk.red.bold([...configCommandsMap].join(', '))}.`
    )
  }

  if (!defaultCommands) {
    throw new Error(`Couldn't find ${chalk.red.bold('commands')} key in configuration file.`)
  }

  return {
    ...defaultCommands,
    ...commands
  }
}

const getConfigCommand = (key: string, commands?: any): string[] => {
  commands = commands || getConfigCommands()
  const command = findKey(key, commands)

  if (!command) {
    throw new Error(`Couldn't find commands for: ${chalk.red.bold(key)} in commands configuration.`)
  }

  return command
}

export {getConfigCommand, getConfigCommands, isConfigCommandsValid}
