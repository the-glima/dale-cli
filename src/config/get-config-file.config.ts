import chalk from 'chalk'
import {existsSync, readFileSync} from 'fs'

import {Configuration} from '../models/configuration.model'
import {settings} from '../settings'

export const getConfigFile = (configPath = settings.configPath.root): Configuration => {
  let configFile: any

  try {
    existsSync(configPath)
    configFile = readFileSync(configPath, 'utf8')
  } catch {
    try {
      existsSync(settings.configPath.default)
      configFile = readFileSync(settings.configPath.default, 'utf8')
    } catch (error) {
      throw new Error(`Couldn\'t find ${chalk.red.bold(error.path)} file, you need to create one.`)
    }
  }

  return JSON.parse(configFile.toString())
}
