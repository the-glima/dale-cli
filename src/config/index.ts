export * from './get-config-commands.config'
export * from './get-config-file.config'
export * from './get-config-images.config'
export * from './get-config-log-color.config'
export * from './get-config-logo.config'
export * from './get-package-json-name-trimmed.config'
export * from './read-package-json.config'
export * from './replace-config-placeholder.config'
