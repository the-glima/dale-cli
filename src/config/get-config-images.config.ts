import {findKey} from '../utils/find-key.util'

import {getConfigFile} from './get-config-file.config'

export const getConfigImages = (config?: any): string[] | null => {
  config = config || getConfigFile()
  const imageArray = findKey('images', config)
  const imageValue = imageArray && imageArray.length > 0 ? imageArray : null
  const imageFilteredValues = imageValue && imageValue.filter((val: any) => !!val && typeof val === 'string')

  return imageFilteredValues && imageFilteredValues.length ? imageFilteredValues : null
}
