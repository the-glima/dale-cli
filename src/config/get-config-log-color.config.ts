import {getConfigFile} from '../config/get-config-file.config'
import {ConfigLogColor} from '../models/config-log-colors.model'
import {settings} from '../settings'
import {findKey} from '../utils/find-key.util'

export const getConfigLogColor = (config?: any): ConfigLogColor => {
  config = config || getConfigFile()
  const logColor = findKey('logColor', config)

  return logColor ? logColor : settings.logColor
}
