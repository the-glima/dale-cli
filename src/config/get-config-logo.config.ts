import {getConfigFile} from '../config/get-config-file.config'
import {settings} from '../settings'
import {findKey} from '../utils/find-key.util'

export const getConfigLogo = (config?: any): boolean => {
  config = config || getConfigFile()
  const logo = config && findKey('logo', config)

  return logo !== null ? logo : settings.logo
}
