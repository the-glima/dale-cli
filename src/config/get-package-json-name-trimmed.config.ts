import {readPackageJson} from './read-package-json.config'

export const getPackageNameTrimmed = (packageJsonPath?: string): string | null => {
  const regex = /(?!.*\/).+/
  const packageJsonName = readPackageJson('name', packageJsonPath)
  const name = regex.exec(packageJsonName)

  return name && name.length ? name[0] : null
}
