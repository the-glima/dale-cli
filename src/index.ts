import {Command, run} from '@oclif/command'
import chalk from 'chalk'
import {spawnSync} from 'child_process'

import * as config from './config/'
import {CommandOptions} from './models/command-options.model'
import * as settings from './settings'
import * as util from './utils/'

interface CommandArgs {
  name: string
  message?: string
  oclifCommand: Command
  options?: CommandOptions | null
}

const runCommand = ({name, oclifCommand, message, options}: CommandArgs): void => {
  const cmdArgs = util.getCommandArguments(name)
  const command = config.getConfigCommand(name)

  if (!Array.isArray(command)) {
    throw new Error(`Commands for ${chalk.red.bold(name)} should be defined in a array`)
  }

  if (message) util.logCommandMessage(oclifCommand, message)

  command.forEach((cmd: any) => {
    const cmdParsed = config.replaceConfigPlaceholder(cmd, options)
    const cmdReplaced = util.replaceNpmCommand(cmdParsed)

    if (!settings.isTestEnv()) {
      if (util.isArgCommand(name)) {
        oclifCommand.log(`${chalk.gray(cmdReplaced)} ${chalk.gray(...cmdArgs)}\n`)

        spawnSync(cmdReplaced, cmdArgs, {
          stdio: 'inherit',
          shell: true
        })
      } else {
        oclifCommand.log(`${chalk.gray(cmdReplaced)}\n`)

        spawnSync(cmdReplaced, {
          stdio: 'inherit',
          shell: true
        })
      }
    } else {
      oclifCommand.log(`${chalk.gray(cmdReplaced)}\n`)
    }
  })
}

export {run, runCommand}
