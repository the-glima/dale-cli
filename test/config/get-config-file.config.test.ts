import * as path from 'path'

import * as config from '../../src/config/'
import {Configuration} from '../../src/models/configuration.model'
import * as helper from '../helpers'

describe('Config: Get Config File', () => {
  const configDefaultFile: Configuration = helper.getFixtureDefaultConfigFile()
  const configRootFile: Configuration = helper.getFixtureRootConfigFile()

  test('should read and return root parsed configuration', async () => {
    expect(configRootFile).toMatchSnapshot()
  })

  test('should read and return default parsed configuration', async () => {
    expect(configDefaultFile).toMatchSnapshot()
  })

  test("should get default config when can't find the root one", async () => {
    let configDefault: any

    try {
      configDefault = config.getConfigFile(path.resolve(__dirname, 'rootrc.config.json'))
    } catch {}

    expect(configDefault).toMatchSnapshot()
  })

  test("should throw an error if can't find any config file", async () => {
    try {
      helper.getFixtureConfigFile('root-not-found.config.json')
    } catch {
      try {
        helper.getFixtureConfigFile('default-not-found.config.json')
      } catch (error) {
        const errorStringified = error.toString()

        expect(errorStringified).toMatch("Couldn't find")
        expect(errorStringified).toMatch('default-not-found.config.json')
        expect(errorStringified).toMatch('file, you need to create one.')
      }
    }
  })
})
