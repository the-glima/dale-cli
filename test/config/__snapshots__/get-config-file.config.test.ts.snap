// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config: Get Config File should get default config when can't find the root one 1`] = `
Object {
  "commands": Object {
    "build": Array [
      "npx tsc -p tsconfig.build.json",
    ],
    "clean": Array [
      "docker ps -aqf name=<%= IMAGE =%> | xargs -n 1 docker rm",
      "docker images <%= IMAGE =%> -q | xargs -n 1 docker rmi",
      "docker image prune -af",
    ],
    "format": Array [
      "npx prettier --write \\"{src,test}/**/*.ts\\"",
    ],
    "integration": Array [
      "node newman.js",
    ],
    "lint": Array [
      "npx tslint -p tsconfig.json -c tslint.json",
    ],
    "start": Array [
      "docker network ls | grep payvision || docker network create -d bridge payvision",
      "docker-compose -f docker-compose<%= ENV =%>.yaml up",
    ],
    "stop": Array [
      "docker-compose -f docker-compose<%= ENV =%>.yaml down -v --rmi=local",
    ],
    "testcafe": Array [
      "set FRONTEND_URL=<%= URL =%> && npx testcafe --speed 1 chrome ./main.js",
    ],
    "unit": Array [
      "npx jest --runInBand --testPathIgnorePatterns=submodules",
    ],
  },
  "images": Array [
    "foo",
    "bar",
  ],
  "logColor": "cyan",
  "logo": true,
}
`;

exports[`Config: Get Config File should read and return default parsed configuration 1`] = `
Object {
  "commands": Object {
    "clean": Array [
      "echo this is the clean test command from default",
    ],
    "unit": Array [
      "unit -f -env this is the first unit test command from default",
      "unit -ls this is the second unit test command from default",
    ],
  },
  "images": Array [
    "first-image",
    "second-image",
  ],
  "logColor": "magenta",
  "logo": false,
}
`;

exports[`Config: Get Config File should read and return root parsed configuration 1`] = `
Object {
  "commands": Object {
    "reset": Array [
      "echo this is the reset test command from root",
    ],
  },
  "logColor": "red",
  "logo": true,
}
`;
